src/dsl/dsl.jl

```diff
@@ -183,3 +159,21 @@ end
 macro gen(ast::Expr)
     parse_gen_function(ast, Set{Symbol}(), __module__)
 end
+
+macro trace(expr::Expr)
+    return Expr(:gentrace, esc(expr), nothing)
+end
+
+macro trace(expr::Expr, addr)
+    return Expr(:gentrace, esc(expr), Some(addr))
+end
+
+macro param(expr::Expr)
+    if (expr.head != :(::)) error("Syntax in error in @param at $(expr)") end
+    name, type = expr.args
+    return Expr(:genparam, esc(name), esc(type))
+end
+
+macro param(sym::Symbol)
+    return Expr(:genparam, esc(sym), esc(:Any))
```

Comment:
Why `esc(:Any)` rather than `esc(Any)`?  (I would expect `esc(:Any)` to not work because `:Any` is not a type)

Reply:
The `esc` needs to be there, otherwise macro-expansion hygiene automatically adds a `Gen` as a qualifier. With `esc`:
```julia
julia> macroexpand(Main, :(@param theta))
:($(Expr(:genparam, :theta, :Any)))
```
Without `esc` (i.e., `macro param(sym::Symbol) = Expr(:genparam, esc(sym), :Any)`):
```julia
julia> macroexpand(Main, :(@param theta))
:($(Expr(:genparam, :theta, :(Gen.Any))))
```

Reply:
Also, `:genparam` expressions are supposed to take only `Symbol`s as arguments --- it's the job of the of the SML and DML implementations to take those Symbols and appropriately resolve them to types. (At least in the current implementation of the SML parser. The DML just ignores type information for parameters.)